cli/repair/clusterwide_config.go

```diff
@@ -0,0 +1,302 @@
+	topologyConfFilename = "topology.yml"
+
+	keyInstances            = "servers"
+	keyReplicasets          = "replicasets"
+	keyInstanceAdvertiseURI = "uri"
+	keyInstanceDisabled     = "disabled"
+
+	keyInstanceReplicasetUUID = "replicaset_uuid"
+	keyReplicasetLeaders      = "master"
+	keyReplicasetAlias        = "alias"
+	keyReplicasetRoles        = "roles"
+
+	expelledState          = "expelled"
+	unnamedReplicasetAlias = "unnamed"
+)
+
+type RawConfType map[interface{}]interface{}
+
+type InstanceConfType struct {
+	AdvertiseURI   string
+	ReplicasetUUID string
+
+	IsExpelled bool
+	IsDisabled bool
+
+	Raw RawConfType
+}
+
+type ReplicasetConfType struct {
+	Alias     string
+	Instances map[string]bool
+	Leaders   []string
+	Roles     []string
+
+	Raw RawConfType
+}
+
+type TopologyConfType struct {
+	Path    string
+	Content []byte
+
+	Raw RawConfType
+
+	InstancesRaw RawConfType
+	Instances    map[string]InstanceConfType
+
+	ReplicasetsRaw RawConfType
+	Replicasets    map[string]ReplicasetConfType
+}
+
+func getTopologyConfPath(workDir string) string {
+	return filepath.Join(workDir, configDirName, topologyConfFilename)
+}
+
+func getTopologyConf(workDir string) (*TopologyConfType, error) {
+	var err error
+	var topologyConf TopologyConfType
+
+	if fileInfo, err := os.Stat(workDir); err != nil {
+		return nil, fmt.Errorf("Failed to use instance workdir: %s", err)
+	} else if !fileInfo.IsDir() {
+		return nil, fmt.Errorf("%s is not a directory", workDir)
+	}
+
+	topologyConf.Path = getTopologyConfPath(workDir)
+	if _, err := os.Stat(topologyConf.Path); err != nil {
+		return nil, fmt.Errorf("Failed to use topology config path: %s", err)
+	}
+
+	topologyConf.Content, err = common.GetFileContentBytes(topologyConf.Path)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read config: %s", err)
+	}
+
+	if err := yaml.Unmarshal(topologyConf.Content, &topologyConf.Raw); err != nil {
+		return nil, fmt.Errorf("Failed to parse config: %s", err)
+	}
+
+	instancesConf, err := getInstancesConf(&topologyConf)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get instances config: %s", err)
+	}
+	topologyConf.Instances = *instancesConf
+
+	replicasetsConf, err := getReplicasetsConf(&topologyConf)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get replicasets config: %s", err)
+	}
+	topologyConf.Replicasets = *replicasetsConf
+
+	return &topologyConf, nil
+}
+
+// INSTANCES
+
+func getInstancesConf(topologyConf *TopologyConfType) (*map[string]InstanceConfType, error) {
+	instancesConfRaw, found := topologyConf.Raw[keyInstances]
+	if !found {
+		return nil, fmt.Errorf("Topology config doesn't contain %q key", keyInstances)
+	}
+
+	instancesConfRawMap, ok := instancesConfRaw.(RawConfType)
+	if !ok {
+		return nil, fmt.Errorf("%q value isn't a map", keyInstances)
+	}
+
+	topologyConf.InstancesRaw = instancesConfRawMap
+
+	instancesConf := make(map[string]InstanceConfType)
+
+	for instanceUUIDRaw, instanceConfRaw := range instancesConfRawMap {
```

Comment:
Do you validate correctness of specified UUID/URI? Btw null uuid (00000000-0000-0000-0000-000000000000) should be banned as well.

Reply:
It's not `cartridge-cli`'s business. It's about clusterwide config format.

Reply:
You give an instrument to patch configuration by hands. You should prevent such invalid/dangerous cases

Reply:
If somebody specify null uuid, cli raises an error `no instance with such UUID`. 
The only one non-handled case is: cluster already contains instance with non-correct UUID. This is not `cartridge-cli` business.

There is no possibility to change instance UUID, we only operate existent UUIDs, and we can only validate if cluster contains instance/replicaset with such UUID.

Reply:
Well, ok